src/mono/mono/metadata/assembly.c

```diff
@@ -3188,6 +3204,164 @@ mono_register_bundled_satellite_assemblies (const MonoBundledSatelliteAssembly *
 	satellite_bundles = assemblies;
 }
 
+static GHashTable *bundled_resources = NULL;
+
+static int *resources_bundled_counts = NULL;
+
+//---------------------------------------------------------------------------------------
+//
+// mono_add_bundled_resource handles bundling of many types of resources to circumvent
+// needing to find or have those resources on disk. The BundledResource struct models
+// the union of information carried by all supported types of resources which are
+// enumerated in BundledResourceType.
+//
+// bundled_resources:
+// A single hash table will hold all resources being bundled with the understanding
+// that all resources being bundled are uniquely named. The bundled resource is tagged
+// with the type of resource it represents for later processing of specific types of
+// resources as done in mono_register_bundled_resources.
+//
+// resources_bundled_counts:
+// An integer array tracks the number of each resource type added to the bundle
+// to facilitate appropriate memory allocation when processing the bundle as done in
+// mono_register_bundled_resources
+//
+// Arguments:
+//  * name - Unique name of the resource
+//  * culture - Culture of satellite assemblies
+//  * data - Byte data array of the bundled resource
+//  * size - Size of byte data array of the bundled resource
+//  type - Type of bundled resource
+//
+
+void
+mono_add_bundled_resource (const char *name, const char *culture, const unsigned char *data, unsigned int size, BundledResourceType type)
+{
+	if (!bundled_resources)
+		bundled_resources = g_hash_table_new (g_str_hash, g_str_equal);
```

Comment:
Right now there seems to be no way to clean up all allocated resources around bundles, even if we won't call it from library builder at the moment (since it doesn't have a shutdown routine), maybe we should at least have a method that makes sure all gets cleaned up and then it can be called, either by embedder or potentially, runtime during shutdown.

Reply:
Right, I was wondering at what point would it be okay to clear up all the allocated memory for bundled resources. Would that only be during shutdown? Or when the calling embedder knows its done with the mono runtime?
In any case, I'll add a function to free the memory allocated.

Reply:
Lets start with a function that makes sure it frees all memory allocated/owned by the bundle API. Once we have that we could decide where it will be called.

Reply:
The content of the files is "code" not "heap data" because it's part of the object files, right ? 
The only memory which could be freed are the filenames and the hashtable structures.

Reply:
Right, I have locally a freeing function that does that. There is also memory held by the `MonoBundledAssembly` structures allocated to register the bundled assembly and satellite assembly, but that would be a different api